app/src/main/java/com/mpieterse/stride/data/repo/concrete/HabitRepositoryImpl.kt

```
@@ -75,31 +81,101 @@ class HabitRepositoryImpl @Inject constructor(
                     )
                 )
             }
-            return created
+            Clogger.d("HabitRepository", "Successfully created habit ${created.id} on server and cached locally${if (finalImageUrl != null) " with image $finalImageUrl" else " (no image)"}")
+            return created.copy(imageUrl = finalImageUrl)
         } catch (e: HttpException) {
+            // For server errors (500, 502, 503, 504), save locally with Pending sync state
+            // This allows the habit to be visible immediately and synced later
+            if (e.code() == 500 || e.code() == 502 || e.code() == 503 || e.code() == 504) {
```

detekt (1.23.8)

[warning] This condition is too complex (4). Defined complexity threshold for conditions is set to '4'

(detekt.complexity.ComplexCondition)